{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def loadsequencing(file):\n",
    "    seq = SeqIO.read(file, 'abi')\n",
    "    \n",
    "    maxq = np.max(seq.letter_annotations['phred_quality'])\n",
    "    rolling = pd.Series(seq.letter_annotations['phred_quality']).rolling(window=20).mean()\n",
    "    start = (rolling > maxq * 0.9).idxmax()\n",
    "    end = (rolling > maxq * 0.9)[::-1].idxmax()\n",
    "    \n",
    "    return seq.seq[start:end], seq, start, end, np.mean(seq.letter_annotations['phred_quality'][start:end])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'seq' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-1575cfc5200c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mletter_annotations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'phred_quality'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mletter_annotations\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'phred_quality'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrolling\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwindow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmean\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxvline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'g'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxvline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'seq' is not defined"
     ]
    }
   ],
   "source": [
    "plt.plot(seq.letter_annotations['phred_quality'])\n",
    "plt.plot(pd.Series(seq.letter_annotations['phred_quality']).rolling(window=20).mean())\n",
    "plt.axvline(x=start, c='g')\n",
    "plt.axvline(x=end, c='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "SYNTH_DIRECTORY = \"../synth1/\"\n",
    "BUILD_DIRECTORY = SYNTH_DIRECTORY + \"builds/build-1/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "validation = []\n",
    "twist = pd.read_csv('../synth1/backup/fragments.csv')\n",
    "\n",
    "fragments = pd.DataFrame({\n",
    "    'Gene': twist['Sequence ID'].str.rsplit('_', n=1).str[0],\n",
    "    'Fragment': twist['Sequence ID'].str.rsplit('_', n=1).str[1],\n",
    "    'Sequence': twist['Insert Sequence']\n",
    "})\n",
    "\n",
    "# Process single gene fragments\n",
    "for i, gene in fragments[~fragments['Gene'].str.contains('_link_')].groupby('Gene'):\n",
    "    frags = []\n",
    "    for seq in gene['Sequence'].str.upper():\n",
    "        # Virtual digest using BbsI\n",
    "        seq = seq[seq.find('GAAGAC')+8:]\n",
    "        seq = seq[:seq.find('GTCTTC')-2]\n",
    "        frags.append(seq)\n",
    "\n",
    "    # Now we have a list of 'digested' fragments; assemble them based on overlaps.\n",
    "    while len(frags) > 1:\n",
    "        frag = frags.pop(0)\n",
    "        for other in frags:\n",
    "            if other[-4:] == frag[:4]:\n",
    "                frag = other[:-4] + frag\n",
    "                frags.remove(other)\n",
    "                break\n",
    "        frags.append(frag)\n",
    "\n",
    "    # Add assembled full sequence to the validation database\n",
    "    # Remove the first and last nucleotides to get the CDS instead of MoClo adapters\n",
    "    validation.append({'Gene': gene['Gene'].iloc[0], 'Sequence': frags[0][1:-1]})\n",
    "\n",
    "# Process multigene fragments\n",
    "for i, gene in fragments[fragments['Gene'].str.contains('_link_')].iterrows():\n",
    "    names = gene['Gene'].split('_link_')\n",
    "\n",
    "    # Virtual digest using BbsI\n",
    "    seq = gene['Sequence'].upper()\n",
    "    seq = seq[seq.find('GAAGAC')+8:]\n",
    "    seq = seq[:seq.find('GTCTTC')-2]\n",
    "    validation.append({'Gene': names[0], 'Sequence': seq[1:-1]})\n",
    "\n",
    "    # We have a multi-gene fragment, digest with BtgZI\n",
    "    seq = gene['Sequence'].upper()\n",
    "    seq = seq[seq.find('GCGATG')+16:]\n",
    "    seq = seq[:seq.find('CATCGC')-10]\n",
    "    validation.append({'Gene': names[1], 'Sequence': seq[1:-1]})\n",
    "    \n",
    "validation_dna = pd.DataFrame(validation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "validation_dna.to_csv(SYNTH_DIRECTORY + \"optimized.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1001_MMSYN1_0115_M13-Forward---20-_A01.ab1\n",
      "Validating MMSYN1_0115 (1-A01)\n",
      "Quality 60.0085034014 60.8630952381\n",
      "Forward matches gene: scores are 542.0 and 542\n",
      "Reverse matches gene: scores are 475.0 and 475\n",
      "1\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1002_MMSYN1_0511_M13-Forward---20-_B01.ab1\n",
      "Validating MMSYN1_0511 (1-B01)\n",
      "Quality 60.3409893993 59.5457685665\n",
      "Forward doesn't match gene: scores are 354.0 and 520\n",
      "Forward aligns to backbone: scores are 566.0 and 566\n",
      "Reverse doesn't match gene: scores are 370.0 and 540\n",
      "Reverse aligns to backbone: scores are 579.0 and 579\n",
      "1\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1003_MMSYN1_0215_M13-Forward---20-_C01.ab1\n",
      "Validating MMSYN1_0215 (1-C01)\n",
      "Quality 60.3360128617 60.423670669\n",
      "Forward doesn't match gene: scores are 313.0 and 572\n",
      "Forward aligns to backbone: scores are 622.0 and 622\n",
      "Reverse doesn't match gene: scores are 311.0 and 540\n",
      "Reverse aligns to backbone: scores are 583.0 and 583\n",
      "2\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1004_MMSYN1_0693_M13-Forward---20-_D01.ab1\n",
      "Validating MMSYN1_0693 (1-D01)\n",
      "Quality 58.6042345277 59.0018315018\n",
      "Forward matches gene: scores are 562.0 and 562\n",
      "Reverse matches gene: scores are 502.0 and 502\n",
      "2\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1005_MMSYN1_0107_M13-Forward---20-_E01.ab1\n",
      "Validating MMSYN1_0107 (1-E01)\n",
      "Quality 59.2717770035 58.25\n",
      "Forward doesn't match gene: scores are 287.0 and 524\n",
      "Forward aligns to backbone: scores are 574.0 and 574\n",
      "Reverse doesn't match gene: scores are 286.0 and 524\n",
      "Reverse aligns to backbone: scores are 564.0 and 564\n",
      "3\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1006_MMSYN1_0148_M13-Forward---20-_F01.ab1\n",
      "Validating MMSYN1_0148 (1-F01)\n",
      "Quality 58.0346083789 57.5088339223\n",
      "Forward doesn't match gene: scores are 292.0 and 500\n",
      "Forward aligns to backbone: scores are 549.0 and 549\n",
      "Reverse doesn't match gene: scores are 303.0 and 525\n",
      "Reverse aligns to backbone: scores are 566.0 and 566\n",
      "4\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1007_MMSYN1_0494_M13-Forward---20-_G01.ab1\n",
      "Validating MMSYN1_0494 (1-G01)\n",
      "Quality 60.5993208829 59.8283464567\n",
      "Forward doesn't match gene: scores are 382.0 and 539\n",
      "Forward aligns to backbone: scores are 589.0 and 589\n",
      "Reverse doesn't match gene: scores are 411.0 and 597\n",
      "Reverse aligns to backbone: scores are 635.0 and 635\n",
      "5\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1008_MMSYN1_0696_M13-Forward---20-_H01.ab1\n",
      "Validating MMSYN1_0696 (1-H01)\n",
      "Quality 60.6040955631 60.2552504039\n",
      "Forward doesn't match gene: scores are 391.0 and 536\n",
      "Forward aligns to backbone: scores are 586.0 and 586\n",
      "Reverse doesn't match gene: scores are 403.0 and 575\n",
      "Reverse aligns to backbone: scores are 619.0 and 619\n",
      "6\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1009_MMSYN1_0530_M13-Forward---20-_A02.ab1\n",
      "Validating MMSYN1_0530 (1-A02)\n",
      "Quality 59.2067226891 58.3322091062\n",
      "Forward doesn't match gene: scores are 359.0 and 545\n",
      "Forward aligns to backbone: scores are 595.0 and 595\n",
      "Reverse doesn't match gene: scores are 365.0 and 550\n",
      "Reverse aligns to backbone: scores are 593.0 and 593\n",
      "7\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1010_MMSYN1_0481_M13-Forward---20-_B02.ab1\n",
      "Validating MMSYN1_0481 (1-B02)\n",
      "Quality 60.2374100719 59.7042735043\n",
      "Forward doesn't match gene: scores are 296.0 and 508\n",
      "Forward aligns to backbone: scores are 556.0 and 556\n",
      "Reverse doesn't match gene: scores are 312.0 and 546\n",
      "Reverse aligns to backbone: scores are 585.0 and 585\n",
      "8\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1011_MMSYN1_0907_M13-Forward---20-_C02.ab1\n",
      "Validating MMSYN1_0907 (1-C02)\n",
      "Quality 60.5031948882 60.6586620926\n",
      "Forward doesn't match gene: scores are 427.0 and 576\n",
      "Forward aligns to backbone: scores are 626.0 and 626\n",
      "Reverse doesn't match gene: scores are 410.0 and 541\n",
      "Reverse aligns to backbone: scores are 583.0 and 583\n",
      "9\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1012_MMSYN1_0281_M13-Forward---20-_D02.ab1\n",
      "Validating MMSYN1_0281 (1-D02)\n",
      "Quality 60.1833910035 60.2904347826\n",
      "Forward doesn't match gene: scores are 372.0 and 531\n",
      "Forward aligns to backbone: scores are 578.0 and 578\n",
      "Reverse doesn't match gene: scores are 376.0 and 535\n",
      "Reverse aligns to backbone: scores are 575.0 and 575\n",
      "10\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1013_MMSYN1_0833_M13-Forward---20-_E02.ab1\n",
      "Validating MMSYN1_0833 (1-E02)\n",
      "Quality 60.1650943396 60.5504761905\n",
      "Forward doesn't match gene: scores are 443.0 and 588\n",
      "Forward aligns to backbone: scores are 636.0 and 636\n",
      "Reverse doesn't match gene: scores are 443.0 and 489\n",
      "Reverse aligns to backbone: scores are 525.0 and 525\n",
      "11\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1014_MMSYN1_0346_M13-Forward---20-_F02.ab1\n",
      "Validating MMSYN1_0346 (1-F02)\n",
      "Quality 60.6404109589 60.0300632911\n",
      "Forward matches gene: scores are 534.0 and 534\n",
      "Reverse matches gene: scores are 586.0 and 586\n",
      "3\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1015_MMSYN1_0421_M13-Forward---20-_G02.ab1\n",
      "Validating MMSYN1_0421 (1-G02)\n",
      "Quality 60.4455284553 13.6787709497\n",
      "Forward matches gene: scores are 312.0 and 564\n",
      "Reverse doesn't match gene: scores are 148.0 and 213\n",
      "Reverse doesn't align to backbone: scores are 290.0 and 358\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1016_MMSYN1_0168_M13-Forward---20-_H02.ab1\n",
      "Validating MMSYN1_0168 (1-H02)\n",
      "Quality 59.9267886855 60.2\n",
      "Forward doesn't match gene: scores are 539.0 and 553\n",
      "Forward aligns to backbone: scores are 601.0 and 601\n",
      "Reverse matches gene: scores are 1.0 and 1\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1017_MMSYN1_0435_M13-Forward---20-_A03.ab1\n",
      "Validating MMSYN1_0435 (1-A03)\n",
      "Quality 60.3280701754 58.3466334165\n",
      "Forward doesn't match gene: scores are 423.0 and 523\n",
      "Forward aligns to backbone: scores are 570.0 and 570\n",
      "Reverse doesn't match gene: scores are 316.0 and 342\n",
      "Reverse aligns to backbone: scores are 401.0 and 401\n",
      "12\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1018_MMSYN1_0253_M13-Forward---20-_B03.ab1\n",
      "Validating MMSYN1_0253 (1-B03)\n",
      "Quality 60.1634615385 60.1305309735\n",
      "Forward doesn't match gene: scores are 332.0 and 572\n",
      "Forward aligns to backbone: scores are 624.0 and 624\n",
      "Reverse doesn't match gene: scores are 302.0 and 450\n",
      "Reverse aligns to backbone: scores are 452.0 and 452\n",
      "13\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1019_MMSYN1_0138_M13-Forward---20-_C03.ab1\n",
      "Validating MMSYN1_0138 (1-C03)\n",
      "Quality 60.1266025641 59.9664991625\n",
      "Forward doesn't match gene: scores are 552.0 and 574\n",
      "Forward aligns to backbone: scores are 624.0 and 624\n",
      "Reverse doesn't match gene: scores are 552.0 and 568\n",
      "Reverse aligns to backbone: scores are 597.0 and 597\n",
      "14\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1020_MMSYN1_0389_M13-Forward---20-_D03.ab1\n",
      "Validating MMSYN1_0389 (1-D03)\n",
      "Quality 60.4654895666 60.6449275362\n",
      "Forward doesn't match gene: scores are 569.0 and 576\n",
      "Forward aligns to backbone: scores are 623.0 and 623\n",
      "Reverse doesn't match gene: scores are 511.0 and 512\n",
      "Reverse aligns to backbone: scores are 552.0 and 552\n",
      "15\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1021_MMSYN1_0039_M13-Forward---20-_E03.ab1\n",
      "Validating MMSYN1_0039 (1-E03)\n",
      "Quality 60.285942492 60.2948275862\n",
      "Forward doesn't match gene: scores are 561.0 and 576\n",
      "Forward aligns to backbone: scores are 626.0 and 626\n",
      "Reverse doesn't match gene: scores are 534.0 and 541\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "16\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1022_MMSYN1_0043_M13-Forward---20-_F03.ab1\n",
      "Validating MMSYN1_0043 (1-F03)\n",
      "Quality 60.285483871 60.6644182125\n",
      "Forward matches gene: scores are 573.0 and 573\n",
      "Reverse matches gene: scores are 548.0 and 548\n",
      "4\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1023_MMSYN1_0447_M13-Forward---20-_G03.ab1\n",
      "Validating MMSYN1_0447 (1-G03)\n",
      "Quality 60.1098726115 60.4135702746\n",
      "Forward doesn't match gene: scores are 336.0 and 576\n",
      "Forward aligns to backbone: scores are 628.0 and 628\n",
      "Reverse doesn't match gene: scores are 340.0 and 575\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reverse aligns to backbone: scores are 619.0 and 619\n",
      "17\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1024_MMSYN1_0045_M13-Forward---20-_H03.ab1\n",
      "Validating MMSYN1_0045 (1-H03)\n",
      "Quality 53.1596858639 60.424137931\n",
      "Forward doesn't match gene: scores are 283.0 and 335\n",
      "Forward aligns to backbone: scores are 382.0 and 382\n",
      "Reverse doesn't match gene: scores are 372.0 and 540\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "18\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1025_MMSYN1_0283_M13-Forward---20-_A04.ab1\n",
      "Validating MMSYN1_0283 (1-A04)\n",
      "Quality 60.498392283 60.1590493601\n",
      "Forward doesn't match gene: scores are 376.0 and 576\n",
      "Forward aligns to backbone: scores are 622.0 and 622\n",
      "Reverse doesn't match gene: scores are 357.0 and 513\n",
      "Reverse aligns to backbone: scores are 547.0 and 547\n",
      "19\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1026_MMSYN1_0416_M13-Forward---20-_B04.ab1\n",
      "Validating MMSYN1_0416 (1-B04)\n",
      "Quality 59.8852988691 60.1032702238\n",
      "Forward doesn't match gene: scores are 291.0 and 572\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 286.0 and 541\n",
      "Reverse aligns to backbone: scores are 581.0 and 581\n",
      "20\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1027_MMSYN1_0839_M13-Forward---20-_C04.ab1\n",
      "Validating MMSYN1_0839 (1-C04)\n",
      "Quality 59.7893890675 60.5017857143\n",
      "Forward doesn't match gene: scores are 265.0 and 576\n",
      "Forward aligns to backbone: scores are 622.0 and 622\n",
      "Reverse doesn't match gene: scores are 249.0 and 531\n",
      "Reverse aligns to backbone: scores are 560.0 and 560\n",
      "21\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1028_MMSYN1_0686_M13-Forward---20-_D04.ab1\n",
      "Validating MMSYN1_0686 (1-D04)\n",
      "Quality 60.2854864434 60.7227036395\n",
      "Forward matches gene: scores are 575.0 and 575\n",
      "Reverse matches gene: scores are 530.0 and 530\n",
      "5\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1029_MMSYN1_0697_M13-Forward---20-_E04.ab1\n",
      "Validating MMSYN1_0697 (1-E04)\n",
      "Quality 55.4460966543 53.8367346939\n",
      "Forward doesn't match gene: scores are 414.0 and 490\n",
      "Forward aligns to backbone: scores are 538.0 and 538\n",
      "Reverse doesn't match gene: scores are 392.0 and 448\n",
      "Reverse aligns to backbone: scores are 490.0 and 490\n",
      "22\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1030_MMSYN1_0286_M13-Forward---20-_F04.ab1\n",
      "Validating MMSYN1_0286 (1-F04)\n",
      "Quality 60.4830053667 60.0831826401\n",
      "Forward doesn't match gene: scores are 398.0 and 512\n",
      "Forward aligns to backbone: scores are 559.0 and 559\n",
      "Reverse doesn't match gene: scores are 409.0 and 541\n",
      "Reverse aligns to backbone: scores are 553.0 and 553\n",
      "23\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1031_MMSYN1_0166_M13-Forward---20-_G04.ab1\n",
      "Validating MMSYN1_0166 (1-G04)\n",
      "Quality 60.0841423948 60.1899827288\n",
      "Forward doesn't match gene: scores are 466.0 and 571\n",
      "Forward aligns to backbone: scores are 618.0 and 618\n",
      "Reverse doesn't match gene: scores are 450.0 and 539\n",
      "Reverse aligns to backbone: scores are 579.0 and 579\n",
      "24\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1032_MMSYN1_0008_M13-Forward---20-_H04.ab1\n",
      "Validating MMSYN1_0008 (1-H04)\n",
      "Quality 60.4390243902 60.4148020654\n",
      "Forward doesn't match gene: scores are 437.0 and 523\n",
      "Forward aligns to backbone: scores are 574.0 and 574\n",
      "Reverse doesn't match gene: scores are 437.0 and 541\n",
      "Reverse aligns to backbone: scores are 581.0 and 581\n",
      "25\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1033_MMSYN1_0838_M13-Forward---20-_A05.ab1\n",
      "Validating MMSYN1_0838 (1-A05)\n",
      "Quality 60.0779014308 60.1155234657\n",
      "Forward doesn't match gene: scores are 416.0 and 582\n",
      "Forward aligns to backbone: scores are 629.0 and 629\n",
      "Reverse doesn't match gene: scores are 384.0 and 513\n",
      "Reverse aligns to backbone: scores are 554.0 and 554\n",
      "26\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1034_MMSYN1_0433_M13-Forward---20-_B05.ab1\n",
      "Validating MMSYN1_0433 (1-B05)\n",
      "Quality 53.0378787879 56.7211328976\n",
      "Forward doesn't match gene: scores are 210.0 and 217\n",
      "Forward aligns to backbone: scores are 264.0 and 264\n",
      "Reverse doesn't match gene: scores are 358.0 and 453\n",
      "Reverse aligns to backbone: scores are 459.0 and 459\n",
      "27\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1035_MMSYN1_0066_M13-Forward---20-_C05.ab1\n",
      "Validating MMSYN1_0066 (1-C05)\n",
      "Quality 60.0826833073 60.0923824959\n",
      "Forward matches gene: scores are 591.0 and 591\n",
      "Reverse matches gene: scores are 570.0 and 570\n",
      "6\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1036_MMSYN1_0885_M13-Forward---20-_D05.ab1\n",
      "Validating MMSYN1_0885 (1-D05)\n",
      "Quality 60.2930153322 60.2104316547\n",
      "Forward doesn't match gene: scores are 562.0 and 571\n",
      "Forward aligns to backbone: scores are 587.0 and 587\n",
      "Reverse doesn't match gene: scores are 515.0 and 518\n",
      "Reverse aligns to backbone: scores are 556.0 and 556\n",
      "28\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1037_MMSYN1_0144_M13-Forward---20-_E05.ab1\n",
      "Validating MMSYN1_0144 (1-E05)\n",
      "Quality 58.9306260575 56.9593495935\n",
      "Forward doesn't match gene: scores are 329.0 and 547\n",
      "Forward aligns to backbone: scores are 591.0 and 591\n",
      "Reverse doesn't match gene: scores are 314.0 and 491\n",
      "Reverse aligns to backbone: scores are 492.0 and 492\n",
      "29\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1038_MMSYN1_0326_M13-Forward---20-_F05.ab1\n",
      "Validating MMSYN1_0326 (1-F05)\n",
      "Quality 60.2256 60.6117850953\n",
      "Forward matches gene: scores are 578.0 and 578\n",
      "Reverse matches gene: scores are 538.0 and 538\n",
      "7\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1039_MMSYN1_0327_M13-Forward---20-_G05.ab1\n",
      "Validating MMSYN1_0327 (1-G05)\n",
      "Quality 60.1207729469 58.8525252525\n",
      "Forward doesn't match gene: scores are 433.0 and 574\n",
      "Forward aligns to backbone: scores are 621.0 and 621\n",
      "Reverse doesn't match gene: scores are 392.0 and 487\n",
      "Reverse aligns to backbone: scores are 495.0 and 495\n",
      "30\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1040_MMSYN1_0097_M13-Forward---20-_H05.ab1\n",
      "Validating MMSYN1_0097 (1-H05)\n",
      "Quality 60.4077181208 59.8448275862\n",
      "Forward doesn't match gene: scores are 436.0 and 545\n",
      "Forward aligns to backbone: scores are 596.0 and 596\n",
      "Reverse doesn't match gene: scores are 435.0 and 539\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "31\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1041_MMSYN1_0108_M13-Forward---20-_A06.ab1\n",
      "Validating MMSYN1_0108 (1-A06)\n",
      "Quality 59.9069020867 60.4555765595\n",
      "Forward doesn't match gene: scores are 479.0 and 576\n",
      "Forward aligns to backbone: scores are 623.0 and 623\n",
      "Reverse doesn't match gene: scores are 448.0 and 514\n",
      "Reverse aligns to backbone: scores are 529.0 and 529\n",
      "32\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1042_MMSYN1_0503_M13-Forward---20-_B06.ab1\n",
      "Validating MMSYN1_0503 (1-B06)\n",
      "Quality 60.4222222222 60.2714025501\n",
      "Forward doesn't match gene: scores are 299.0 and 583\n",
      "Forward aligns to backbone: scores are 630.0 and 630\n",
      "Reverse doesn't match gene: scores are 290.0 and 513\n",
      "Reverse aligns to backbone: scores are 549.0 and 549\n",
      "33\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1043_MMSYN1_0077_M13-Forward---20-_C06.ab1\n",
      "Validating MMSYN1_0077 (1-C06)\n",
      "Quality 60.151799687 60.5487804878\n",
      "Forward matches gene: scores are 592.0 and 592\n",
      "Reverse matches gene: scores are 530.0 and 530\n",
      "8\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1044_MMSYN1_0859_M13-Forward---20-_D06.ab1\n",
      "Validating MMSYN1_0859 (1-D06)\n",
      "Quality 59.1177446103 57.5010526316\n",
      "Forward doesn't match gene: scores are 564.0 and 568\n",
      "Forward aligns to backbone: scores are 603.0 and 603\n",
      "Reverse matches gene: scores are 446.0 and 446\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1045_MMSYN1_0817_M13-Forward---20-_E06.ab1\n",
      "Validating MMSYN1_0817 (1-E06)\n",
      "Quality 60.2907915994 59.9293103448\n",
      "Forward doesn't match gene: scores are 458.0 and 572\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 440.0 and 540\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "34\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1046_MMSYN1_0853_M13-Forward---20-_F06.ab1\n",
      "Validating MMSYN1_0853 (1-F06)\n",
      "Quality 59.8056379822 60.1885098743\n",
      "Forward doesn't match gene: scores are 428.0 and 628\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward aligns to backbone: scores are 674.0 and 674\n",
      "Reverse doesn't match gene: scores are 382.0 and 517\n",
      "Reverse aligns to backbone: scores are 557.0 and 557\n",
      "35\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1047_MMSYN1_0302_M13-Forward---20-_G06.ab1\n",
      "Validating MMSYN1_0302 (1-G06)\n",
      "Quality 59.9919224556 58.6724470135\n",
      "Forward doesn't match gene: scores are 395.0 and 572\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 368.0 and 511\n",
      "Reverse aligns to backbone: scores are 519.0 and 519\n",
      "36\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1048_MMSYN1_0614_M13-Forward---20-_H06.ab1\n",
      "Validating MMSYN1_0614 (1-H06)\n",
      "Quality 60.2656 60.1256454389\n",
      "Forward doesn't match gene: scores are 481.0 and 574\n",
      "Forward aligns to backbone: scores are 625.0 and 625\n",
      "Reverse doesn't match gene: scores are 459.0 and 541\n",
      "Reverse aligns to backbone: scores are 581.0 and 581\n",
      "37\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1049_MMSYN1_0440_M13-Forward---20-_A07.ab1\n",
      "Validating MMSYN1_0440 (1-A07)\n",
      "Quality 59.5190082645 60.2990990991\n",
      "Forward matches gene: scores are 558.0 and 558\n",
      "Reverse matches gene: scores are 512.0 and 512\n",
      "9\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1050_MMSYN1_0388_M13-Forward---20-_B07.ab1\n",
      "Validating MMSYN1_0388 (1-B07)\n",
      "Quality 60.2936630603 60.1894736842\n",
      "Forward doesn't match gene: scores are 382.0 and 600\n",
      "Forward aligns to backbone: scores are 647.0 and 647\n",
      "Reverse doesn't match gene: scores are 363.0 and 541\n",
      "Reverse aligns to backbone: scores are 570.0 and 570\n",
      "38\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1051_MMSYN1_0906_M13-Forward---20-_C07.ab1\n",
      "Validating MMSYN1_0906 (1-C07)\n",
      "Quality 60.6557377049 60.8262350937\n",
      "Forward matches gene: scores are 563.0 and 563\n",
      "Reverse matches gene: scores are 548.0 and 548\n",
      "10\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1052_MMSYN1_0264_M13-Forward---20-_D07.ab1\n",
      "Validating MMSYN1_0264 (1-D07)\n",
      "Quality 60.1286644951 60.1094890511\n",
      "Forward doesn't match gene: scores are 473.0 and 570\n",
      "Forward aligns to backbone: scores are 614.0 and 614\n",
      "Reverse doesn't match gene: scores are 441.0 and 509\n",
      "Reverse aligns to backbone: scores are 548.0 and 548\n",
      "39\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1053_MMSYN1_0352_M13-Forward---20-_E07.ab1\n",
      "Validating MMSYN1_0352 (1-E07)\n",
      "Quality 60.3017656501 60.4894366197\n",
      "Forward doesn't match gene: scores are 355.0 and 576\n",
      "Forward aligns to backbone: scores are 623.0 and 623\n",
      "Reverse doesn't match gene: scores are 341.0 and 541\n",
      "Reverse aligns to backbone: scores are 568.0 and 568\n",
      "40\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1054_MMSYN1_0777_M13-Forward---20-_F07.ab1\n",
      "Validating MMSYN1_0777 (1-F07)\n",
      "Quality 60.0823909532 59.3902027027\n",
      "Forward doesn't match gene: scores are 338.0 and 575\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 334.0 and 563\n",
      "Reverse aligns to backbone: scores are 592.0 and 592\n",
      "41\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1055_MMSYN1_0167_M13-Forward---20-_G07.ab1\n",
      "Validating MMSYN1_0167 (1-G07)\n",
      "Quality 60.2326332795 60.2025316456\n",
      "Forward doesn't match gene: scores are 546.0 and 572\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 500.0 and 513\n",
      "Reverse aligns to backbone: scores are 553.0 and 553\n",
      "42\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1056_MMSYN1_0505_M13-Forward---20-_H07.ab1\n",
      "Validating MMSYN1_0505 (1-H07)\n",
      "Quality 60.1107544141 59.9786407767\n",
      "Forward doesn't match gene: scores are 474.0 and 574\n",
      "Forward aligns to backbone: scores are 623.0 and 623\n",
      "Reverse doesn't match gene: scores are 435.0 and 494\n",
      "Reverse aligns to backbone: scores are 515.0 and 515\n",
      "43\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1057_MMSYN1_0220_M13-Forward---20-_A08.ab1\n",
      "Validating MMSYN1_0220 (1-A08)\n",
      "Quality 60.4345047923 59.9064516129\n",
      "Forward matches gene: scores are 576.0 and 576\n",
      "Reverse matches gene: scores are 579.0 and 579\n",
      "11\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1058_MMSYN1_0672_M13-Forward---20-_B08.ab1\n",
      "Validating MMSYN1_0672 (1-B08)\n",
      "Quality 59.9433656958 59.8821362799\n",
      "Forward doesn't match gene: scores are 257.0 and 568\n",
      "Forward aligns to backbone: scores are 618.0 and 618\n",
      "Reverse doesn't match gene: scores are 251.0 and 517\n",
      "Reverse aligns to backbone: scores are 543.0 and 543\n",
      "44\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1059_MMSYN1_0372_M13-Forward---20-_C08.ab1\n",
      "Validating MMSYN1_0372 (1-C08)\n",
      "Quality 58.1729200653 59.5217391304\n",
      "Forward doesn't match gene: scores are 553.0 and 567\n",
      "Forward aligns to backbone: scores are 613.0 and 613\n",
      "Reverse doesn't match gene: scores are 511.0 and 518\n",
      "Reverse aligns to backbone: scores are 529.0 and 529\n",
      "45\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1060_MMSYN1_0296_M13-Forward---20-_D08.ab1\n",
      "Validating MMSYN1_0296 (1-D08)\n",
      "Quality 59.281098546 58.1078998073\n",
      "Forward doesn't match gene: scores are 374.0 and 572\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 354.0 and 515\n",
      "Reverse aligns to backbone: scores are 519.0 and 519\n",
      "46\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1061_MMSYN1_0054_M13-Forward---20-_E08.ab1\n",
      "Validating MMSYN1_0054 (1-E08)\n",
      "Quality 59.1404958678 57.7198067633\n",
      "Forward doesn't match gene: scores are 321.0 and 571\n",
      "Forward aligns to backbone: scores are 605.0 and 605\n",
      "Reverse doesn't match gene: scores are 273.0 and 412\n",
      "Reverse aligns to backbone: scores are 414.0 and 414\n",
      "47\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1062_MMSYN1_0298_M13-Forward---20-_F08.ab1\n",
      "Validating MMSYN1_0298 (1-F08)\n",
      "Quality 60.1851851852 56.5401301518\n",
      "Forward doesn't match gene: scores are 245.0 and 550\n",
      "Forward aligns to backbone: scores are 594.0 and 594\n",
      "Reverse doesn't match gene: scores are 224.0 and 438\n",
      "Reverse aligns to backbone: scores are 461.0 and 461\n",
      "48\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1063_MMSYN1_0139_M13-Forward---20-_G08.ab1\n",
      "Validating MMSYN1_0139 (1-G08)\n",
      "Quality 59.4407079646 60.9285714286\n",
      "Forward doesn't match gene: scores are 528.0 and 529\n",
      "Forward aligns to backbone: scores are 565.0 and 565\n",
      "Reverse matches gene: scores are 520.0 and 520\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1064_MMSYN1_0270_M13-Forward---20-_H08.ab1\n",
      "Validating MMSYN1_0270 (1-H08)\n",
      "Quality 60.1227786753 59.9982110912\n",
      "Forward doesn't match gene: scores are 302.0 and 573\n",
      "Forward aligns to backbone: scores are 619.0 and 619\n",
      "Reverse doesn't match gene: scores are 291.0 and 520\n",
      "Reverse aligns to backbone: scores are 559.0 and 559\n",
      "49\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1065_MMSYN1_0371_M13-Forward---20-_A09.ab1\n",
      "Validating MMSYN1_0371 (1-A09)\n",
      "Quality 60.5224913495 59.9036144578\n",
      "Forward doesn't match gene: scores are 517.0 and 527\n",
      "Forward aligns to backbone: scores are 578.0 and 578\n",
      "Reverse doesn't match gene: scores are 530.0 and 541\n",
      "Reverse aligns to backbone: scores are 581.0 and 581\n",
      "50\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1066_MMSYN1_0042_M13-Forward---20-_B09.ab1\n",
      "Validating MMSYN1_0042 (1-B09)\n",
      "Quality 60.4581993569 53.0899470899\n",
      "Forward doesn't match gene: scores are 440.0 and 574\n",
      "Forward aligns to backbone: scores are 622.0 and 622\n",
      "Reverse matches gene: scores are 101.0 and 101\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1067_MMSYN1_0132_M13-Forward---20-_C09.ab1\n",
      "Validating MMSYN1_0132 (1-C09)\n",
      "Quality 60.249201278 60.6462585034\n",
      "Forward doesn't match gene: scores are 474.0 and 576\n",
      "Forward aligns to backbone: scores are 626.0 and 626\n",
      "Reverse doesn't match gene: scores are 466.0 and 543\n",
      "Reverse aligns to backbone: scores are 588.0 and 588\n",
      "51\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1068_MMSYN1_0525_M13-Forward---20-_D09.ab1\n",
      "Validating MMSYN1_0525 (1-D09)\n",
      "Quality 58.8102836879 55.5428571429\n",
      "Forward doesn't match gene: scores are 288.0 and 523\n",
      "Forward aligns to backbone: scores are 564.0 and 564\n",
      "Reverse doesn't match gene: scores are 269.0 and 444\n",
      "Reverse aligns to backbone: scores are 455.0 and 455\n",
      "52\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1069_MMSYN1_0806_M13-Forward---20-_E09.ab1\n",
      "Validating MMSYN1_0806 (1-E09)\n",
      "Quality 58.5348837209 59.398467433\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward doesn't match gene: scores are 279.0 and 521\n",
      "Forward aligns to backbone: scores are 559.0 and 559\n",
      "Reverse doesn't match gene: scores are 284.0 and 517\n",
      "Reverse aligns to backbone: scores are 522.0 and 522\n",
      "53\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1070_MMSYN1_0338_M13-Forward---20-_F09.ab1\n",
      "Validating MMSYN1_0338 (1-F09)\n",
      "Quality 59.5445859873 58.1513292434\n",
      "Forward doesn't match gene: scores are 409.0 and 576\n",
      "Forward aligns to backbone: scores are 628.0 and 628\n",
      "Reverse doesn't match gene: scores are 357.0 and 460\n",
      "Reverse aligns to backbone: scores are 489.0 and 489\n",
      "54\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1071_MMSYN1_0164_M13-Forward---20-_G09.ab1\n",
      "Validating MMSYN1_0164 (1-G09)\n",
      "Quality 60.4520325203 60.4493996569\n",
      "Forward matches gene: scores are 495.0 and 563\n",
      "Reverse matches gene: scores are 495.0 and 536\n",
      "12\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1072_MMSYN1_0409_M13-Forward---20-_H09.ab1\n",
      "Validating MMSYN1_0409 (1-H09)\n",
      "Quality 60.0961538462 59.8793103448\n",
      "Forward doesn't match gene: scores are 419.0 and 574\n",
      "Forward aligns to backbone: scores are 624.0 and 624\n",
      "Reverse doesn't match gene: scores are 395.0 and 540\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "55\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1073_MMSYN1_0632_M13-Forward---20-_A10.ab1\n",
      "Validating MMSYN1_0632 (1-A10)\n",
      "Quality 60.1379928315 60.3447684391\n",
      "Forward doesn't match gene: scores are 305.0 and 509\n",
      "Forward aligns to backbone: scores are 558.0 and 558\n",
      "Reverse doesn't match gene: scores are 318.0 and 542\n",
      "Reverse aligns to backbone: scores are 583.0 and 583\n",
      "56\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1074_MMSYN1_0772_M13-Forward---20-_B10.ab1\n",
      "Validating MMSYN1_0772 (1-B10)\n",
      "Quality 60.2798742138 59.5837651123\n",
      "Forward doesn't match gene: scores are 333.0 and 585\n",
      "Forward aligns to backbone: scores are 636.0 and 636\n",
      "Reverse doesn't match gene: scores are 314.0 and 539\n",
      "Reverse aligns to backbone: scores are 579.0 and 579\n",
      "57\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1075_MMSYN1_0881_M13-Forward---20-_C10.ab1\n",
      "Validating MMSYN1_0881 (1-C10)\n",
      "Quality 59.9936 59.7567084079\n",
      "Forward doesn't match gene: scores are 528.0 and 576\n",
      "Forward aligns to backbone: scores are 625.0 and 625\n",
      "Reverse doesn't match gene: scores are 485.0 and 518\n",
      "Reverse aligns to backbone: scores are 559.0 and 559\n",
      "58\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1076_MMSYN1_0401_M13-Forward---20-_D10.ab1\n",
      "Validating MMSYN1_0401 (1-D10)\n",
      "Quality 59.8413461538 60.0051724138\n",
      "Forward doesn't match gene: scores are 539.0 and 576\n",
      "Forward aligns to backbone: scores are 624.0 and 624\n",
      "Reverse doesn't match gene: scores are 520.0 and 540\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "59\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1077_MMSYN1_0909_M13-Forward---20-_E10.ab1\n",
      "Validating MMSYN1_0909 (1-E10)\n",
      "Quality 60.3204508857 57.6821862348\n",
      "Forward doesn't match gene: scores are 269.0 and 574\n",
      "Forward aligns to backbone: scores are 621.0 and 621\n",
      "Reverse doesn't match gene: scores are 247.0 and 468\n",
      "Reverse aligns to backbone: scores are 494.0 and 494\n",
      "60\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1078_MMSYN1_0154_M13-Forward---20-_F10.ab1\n",
      "Validating MMSYN1_0154 (1-F10)\n",
      "Quality 60.2362459547 60.6867256637\n",
      "Forward doesn't match gene: scores are 516.0 and 569\n",
      "Forward aligns to backbone: scores are 618.0 and 618\n",
      "Reverse doesn't match gene: scores are 482.0 and 518\n",
      "Reverse aligns to backbone: scores are 565.0 and 565\n",
      "61\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1079_MMSYN1_0063_M13-Forward---20-_G10.ab1\n",
      "Validating MMSYN1_0063 (1-G10)\n",
      "Quality 60.1587301587 60.3808724832\n",
      "Forward matches gene: scores are 584.0 and 584\n",
      "Reverse matches gene: scores are 549.0 and 549\n",
      "13\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1080_MMSYN1_0109_M13-Forward---20-_H10.ab1\n",
      "Validating MMSYN1_0109 (1-H10)\n",
      "Quality 59.9181380417 58.1234256927\n",
      "Forward doesn't match gene: scores are 443.0 and 571\n",
      "Forward aligns to backbone: scores are 623.0 and 623\n",
      "Reverse doesn't match gene: scores are 348.0 and 391\n",
      "Reverse aligns to backbone: scores are 397.0 and 397\n",
      "62\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1081_MMSYN1_0314_M13-Forward---20-_A11.ab1\n",
      "Validating MMSYN1_0314 (1-A11)\n",
      "Quality 60.3784219002 60.3595317726\n",
      "Forward matches gene: scores are 571.0 and 571\n",
      "Reverse matches gene: scores are 554.0 and 554\n",
      "14\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1082_MMSYN1_0197_M13-Forward---20-_B11.ab1\n",
      "Validating MMSYN1_0197 (1-B11)\n",
      "Quality 60.4806338028 60.5965811966\n",
      "Forward doesn't match gene: scores are 448.0 and 521\n",
      "Forward aligns to backbone: scores are 568.0 and 568\n",
      "Reverse doesn't match gene: scores are 462.0 and 541\n",
      "Reverse aligns to backbone: scores are 585.0 and 585\n",
      "63\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1083_MMSYN1_0438_M13-Forward---20-_C11.ab1\n",
      "Validating MMSYN1_0438 (1-C11)\n",
      "Quality 60.395229983 59.4329738059\n",
      "Forward matches gene: scores are 399.0 and 537\n",
      "Reverse matches gene: scores are 399.0 and 603\n",
      "15\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1084_MMSYN1_0332_M13-Forward---20-_D11.ab1\n",
      "Validating MMSYN1_0332 (1-D11)\n",
      "Quality 60.5233494364 59.6810344828\n",
      "Forward doesn't match gene: scores are 430.0 and 574\n",
      "Forward aligns to backbone: scores are 621.0 and 621\n",
      "Reverse doesn't match gene: scores are 405.0 and 540\n",
      "Reverse aligns to backbone: scores are 580.0 and 580\n",
      "64\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1085_MMSYN1_0430_M13-Forward---20-_E11.ab1\n",
      "Validating MMSYN1_0430 (1-E11)\n",
      "Quality 60.0862619808 60.4733218589\n",
      "Forward doesn't match gene: scores are 267.0 and 576\n",
      "Forward aligns to backbone: scores are 626.0 and 626\n",
      "Reverse doesn't match gene: scores are 269.0 and 540\n",
      "Reverse aligns to backbone: scores are 581.0 and 581\n",
      "65\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1086_MMSYN1_0353_M13-Forward---20-_F11.ab1\n",
      "Validating MMSYN1_0353 (1-F11)\n",
      "Quality 60.0446570973 59.9595070423\n",
      "Forward doesn't match gene: scores are 290.0 and 576\n",
      "Forward aligns to backbone: scores are 627.0 and 627\n",
      "Reverse doesn't match gene: scores are 284.0 and 528\n",
      "Reverse aligns to backbone: scores are 568.0 and 568\n",
      "66\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1087_MMSYN1_0375_M13-Forward---20-_G11.ab1\n",
      "Validating MMSYN1_0375 (1-G11)\n",
      "Quality 60.4891696751 59.5385996409\n",
      "Forward doesn't match gene: scores are 380.0 and 505\n",
      "Forward aligns to backbone: scores are 554.0 and 554\n",
      "Reverse doesn't match gene: scores are 377.0 and 517\n",
      "Reverse aligns to backbone: scores are 557.0 and 557\n",
      "67\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1088_MMSYN1_0730_M13-Forward---20-_H11.ab1\n",
      "Validating MMSYN1_0730 (1-H11)\n",
      "Quality 60.0257352941 60.1717902351\n",
      "Forward doesn't match gene: scores are 346.0 and 497\n",
      "Forward aligns to backbone: scores are 544.0 and 544\n",
      "Reverse doesn't match gene: scores are 355.0 and 513\n",
      "Reverse aligns to backbone: scores are 553.0 and 553\n",
      "68\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1089_MMSYN1_0222_M13-Forward---20-_A12.ab1\n",
      "Validating MMSYN1_0222 (1-A12)\n",
      "Quality 60.6589403974 60.3474714519\n",
      "Forward doesn't match gene: scores are 546.0 and 553\n",
      "Forward aligns to backbone: scores are 604.0 and 604\n",
      "Reverse doesn't match gene: scores are 566.0 and 574\n",
      "Reverse aligns to backbone: scores are 613.0 and 613\n",
      "69\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1090_MMSYN1_0526_link_MMSYN1_0499_M13-Forward---20-_B12.ab1\n",
      "Validating MMSYN1_0526_link_MMSYN1_0499 (1-B12)\n",
      "Quality 59.8971061093 60.361774744\n",
      "Forward doesn't match gene: scores are 174.0 and 572\n",
      "Forward aligns to backbone: scores are 622.0 and 622\n",
      "Reverse doesn't match gene: scores are 176.0 and 541\n",
      "Reverse aligns to backbone: scores are 586.0 and 586\n",
      "70\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1091_MMSYN1_0852_link_MMSYN1_0778_M13-Forward---20-_C12.ab1\n",
      "Validating MMSYN1_0852_link_MMSYN1_0778 (1-C12)\n",
      "Quality 60.0701754386 60.6283783784\n",
      "Forward doesn't match gene: scores are 213.0 and 576\n",
      "Forward aligns to backbone: scores are 627.0 and 627\n",
      "Reverse doesn't match gene: scores are 205.0 and 552\n",
      "Reverse aligns to backbone: scores are 592.0 and 592\n",
      "71\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1092_MMSYN1_0400_M13-Forward---20-_D12.ab1\n",
      "Validating MMSYN1_0400 (1-D12)\n",
      "Quality 60.1485623003 60.202970297\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward doesn't match gene: scores are 359.0 and 574\n",
      "Forward aligns to backbone: scores are 626.0 and 626\n",
      "Reverse doesn't match gene: scores are 355.0 and 566\n",
      "Reverse aligns to backbone: scores are 606.0 and 606\n",
      "72\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1093_MMSYN1_0408_M13-Forward---20-_E12.ab1\n",
      "Validating MMSYN1_0408 (1-E12)\n",
      "Quality 60.1952 60.5211505922\n",
      "Forward doesn't match gene: scores are 394.0 and 574\n",
      "Forward aligns to backbone: scores are 625.0 and 625\n",
      "Reverse doesn't match gene: scores are 381.0 and 543\n",
      "Reverse aligns to backbone: scores are 591.0 and 591\n",
      "73\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1094_MMSYN1_0094_M13-Forward---20-_F12.ab1\n",
      "Validating MMSYN1_0094 (1-F12)\n",
      "Quality 59.9255121043 59.4353146853\n",
      "Forward doesn't match gene: scores are 373.0 and 492\n",
      "Forward aligns to backbone: scores are 537.0 and 537\n",
      "Reverse doesn't match gene: scores are 392.0 and 539\n",
      "Reverse aligns to backbone: scores are 572.0 and 572\n",
      "74\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1095_MMSYN1_0376_M13-Forward---20-_G12.ab1\n",
      "Validating MMSYN1_0376 (1-G12)\n",
      "Quality 60.6466302368 60.0745233969\n",
      "Forward doesn't match gene: scores are 244.0 and 499\n",
      "Forward aligns to backbone: scores are 549.0 and 549\n",
      "Reverse doesn't match gene: scores are 256.0 and 526\n",
      "Reverse aligns to backbone: scores are 577.0 and 577\n",
      "75\n",
      "\n",
      "Analysing sequence ../synth1/builds/build-1/sequencing/CM_525055-1096_MMSYN1_0060_M13-Forward---20-_H12.ab1\n",
      "Validating MMSYN1_0060 (1-H12)\n",
      "Quality 60.1294498382 60.3264957265\n",
      "Forward doesn't match gene: scores are 432.0 and 571\n",
      "Forward aligns to backbone: scores are 618.0 and 618\n",
      "Reverse doesn't match gene: scores are 412.0 and 541\n",
      "Reverse aligns to backbone: scores are 585.0 and 585\n",
      "76\n",
      "\n",
      "good 15\n",
      "rfp 76\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "157"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "import os\n",
    "import re\n",
    "\n",
    "from Bio import pairwise2\n",
    "from Bio import AlignIO\n",
    "from Bio import Align\n",
    "from Bio import Seq\n",
    "from Bio import SeqRecord\n",
    "from Bio.Alphabet import generic_dna\n",
    "\n",
    "forward_primer = \"M13-Forward---20-\"\n",
    "reverse_primer = \"M13-Reverse\"\n",
    "\n",
    "all_alignments = []\n",
    "bad_alignments = []\n",
    "good_name = []\n",
    "good_well = []\n",
    "\n",
    "rfp_name = []\n",
    "rfp_well = []\n",
    "\n",
    "good = 0\n",
    "rfp = 0\n",
    "\n",
    "backbone_seq = SeqIO.read('./popen_v1-1_backbone.fasta', 'fasta')\n",
    "backbone_seq = backbone_seq.seq\n",
    "\n",
    "for seqfile in glob.glob(BUILD_DIRECTORY + \"sequencing/*{}*.ab1\".format(forward_primer)):\n",
    "    \n",
    "    goodF = 0\n",
    "    goodR = 0\n",
    "    rfpF = 0\n",
    "    rfpR = 0\n",
    "\n",
    "    print(\"Analysing sequence\", seqfile)\n",
    "    \n",
    "    initials, order_number, plate_number, well_number, sample_name, primer_name, well_address = re.match(\n",
    "        r'.*/([A-Z]+)_([0-9]+)-([0-9])([0-9]+)_([A-Za-z0-9_-]+)_([A-Za-z0-9-]+)_([A-H][0-9]{2}).ab1',\n",
    "        seqfile).groups()\n",
    "    revfile = \"{}/{}_{}-{}{}_{}_{}_{}.ab1\".format(os.path.dirname(seqfile), initials, order_number, (int(plate_number)+1), well_number, sample_name, reverse_primer, well_address)\n",
    "    \n",
    "    print(\"Validating {} ({}-{})\".format(sample_name, plate_number, well_address))\n",
    "    \n",
    "    # We assume that for every forward there is a reverse\n",
    "    forward, _, _, _, forward_qual = loadsequencing(seqfile)\n",
    "    reverse, revseq, _, _, reverse_qual = loadsequencing(revfile)\n",
    "    \n",
    "    print(\"Quality\", forward_qual, reverse_qual)\n",
    "    \n",
    "#     sample_name = re.sub(r'-*(.*)_[0-9]+', r'\\1', sample_name)\n",
    "    if \"link\" in sample_name:\n",
    "        sample_name = re.sub(r'(.*)_link_.*', r'\\1', sample_name)\n",
    "        \n",
    "    target_seq = validation_dna.loc[validation_dna['Gene']==sample_name, 'Sequence'].iloc[0]\n",
    "        \n",
    "    forward_untrim = forward\n",
    "    reverse_untrim = reverse.reverse_complement()\n",
    "    \n",
    "    forward = forward[forward.find('ATG'):] # Start at start codon\n",
    "    reverse = reverse[reverse.find('TCA'):].reverse_complement() # Stop at stop    \n",
    "    \n",
    "    forward_align = pairwise2.align.globalxx(target_seq, forward, one_alignment_only=True)\n",
    "    reverse_align = pairwise2.align.globalxx(target_seq, reverse, one_alignment_only=True)\n",
    "    \n",
    "    if len(forward_align) == 0 or len(reverse_align) == 0:\n",
    "        if len(forward_align) == 0:\n",
    "            print(\"Forward doesn't align\")\n",
    "\n",
    "        if len(reverse_align) == 0:\n",
    "            print(\"Reverse doesn't align\")\n",
    "            \n",
    "        continue\n",
    "    \n",
    "    forward_align = forward_align[0]\n",
    "    reverse_align = reverse_align[0]\n",
    "    \n",
    "    if (forward_align[2] != min(len(forward), len(target_seq))):\n",
    "        print(\"Forward doesn't match gene: scores are {} and {}\".format(forward_align[2],len(forward)))\n",
    "        forward_align_back = pairwise2.align.globalxx(backbone_seq, forward_untrim, one_alignment_only=True)\n",
    "        forward_align_back = forward_align_back[0]\n",
    "        \n",
    "        if (forward_align_back[2] == min(len(forward_untrim), len(backbone_seq))):\n",
    "            print(\"Forward aligns to backbone: scores are {} and {}\".format(forward_align_back[2],len(forward_untrim)))\n",
    "            rfpF = rfpF + 1\n",
    "        else:\n",
    "            print(\"Forward doesn't align to backbone: scores are {} and {}\".format(forward_align_back[2],len(forward_untrim)))\n",
    "            \n",
    "        align = Align.MultipleSeqAlignment([\n",
    "            SeqRecord.SeqRecord(Seq.Seq(forward_align[0], generic_dna), id=\"TargetSequence\"),\n",
    "            SeqRecord.SeqRecord(Seq.Seq(forward_align[1], generic_dna), id=\"ForwardSequence\"),\n",
    "        ])\n",
    "        bad_alignments.append(align)\n",
    "    else:\n",
    "        print(\"Forward matches gene: scores are {} and {}\".format(forward_align[2],len(forward)))\n",
    "        goodF = goodF + 1\n",
    "        \n",
    "    if (reverse_align[2] != min(len(reverse), len(target_seq))):\n",
    "        print(\"Reverse doesn't match gene: scores are {} and {}\".format(reverse_align[2],len(reverse)))\n",
    "        reverse_align_back = pairwise2.align.globalxx(backbone_seq, reverse_untrim, one_alignment_only=True)\n",
    "        reverse_align_back = reverse_align_back[0]\n",
    "\n",
    "        if (reverse_align_back[2] == min(len(reverse_untrim), len(backbone_seq))):\n",
    "            print(\"Reverse aligns to backbone: scores are {} and {}\".format(reverse_align_back[2],len(reverse_untrim)))\n",
    "            rfpR = 1\n",
    "        else:\n",
    "            print(\"Reverse doesn't align to backbone: scores are {} and {}\".format(reverse_align_back[2],len(reverse_untrim)))            \n",
    "        align = Align.MultipleSeqAlignment([\n",
    "            SeqRecord.SeqRecord(Seq.Seq(reverse_align[0], generic_dna), id=\"TargetSequence\"),\n",
    "            SeqRecord.SeqRecord(Seq.Seq(reverse_align[1], generic_dna), id=\"ReverseSequence\"),\n",
    "        ])\n",
    "        bad_alignments.append(align)\n",
    "    else:\n",
    "        print(\"Reverse matches gene: scores are {} and {}\".format(reverse_align[2],len(reverse)))\n",
    "        goodR = 1\n",
    "    \n",
    "    if goodF + goodR == 2:\n",
    "        good = good + 1\n",
    "        print(good)\n",
    "        good_name.append(sample_name)\n",
    "        good_well.append(well_address)\n",
    "    if rfpF + rfpR == 2:\n",
    "        rfp = rfp + 1\n",
    "        print(rfp)\n",
    "        rfp_name.append(sample_name)\n",
    "        rfp_well.append(well_address)\n",
    "\n",
    "    print()\n",
    "\n",
    "print(\"good\", good)\n",
    "print(\"rfp\",rfp)\n",
    "good = pd.DataFrame({'Gene': good_name,\n",
    "                     'Well': good_well})\n",
    "\n",
    "rfp = pd.DataFrame({'Gene': rfp_name,\n",
    "                    'Well': rfp_well})\n",
    "\n",
    "AlignIO.write(bad_alignments, BUILD_DIRECTORY + 'bad-alignments.aln'.format(initials, order_number), 'clustal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Well</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MMSYN1_0115</td>\n",
       "      <td>A01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MMSYN1_0693</td>\n",
       "      <td>D01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MMSYN1_0346</td>\n",
       "      <td>F02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MMSYN1_0043</td>\n",
       "      <td>F03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MMSYN1_0686</td>\n",
       "      <td>D04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MMSYN1_0066</td>\n",
       "      <td>C05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MMSYN1_0326</td>\n",
       "      <td>F05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MMSYN1_0077</td>\n",
       "      <td>C06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MMSYN1_0440</td>\n",
       "      <td>A07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MMSYN1_0906</td>\n",
       "      <td>C07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MMSYN1_0220</td>\n",
       "      <td>A08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MMSYN1_0164</td>\n",
       "      <td>G09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MMSYN1_0063</td>\n",
       "      <td>G10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MMSYN1_0314</td>\n",
       "      <td>A11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>MMSYN1_0438</td>\n",
       "      <td>C11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Gene Well\n",
       "0   MMSYN1_0115  A01\n",
       "1   MMSYN1_0693  D01\n",
       "2   MMSYN1_0346  F02\n",
       "3   MMSYN1_0043  F03\n",
       "4   MMSYN1_0686  D04\n",
       "5   MMSYN1_0066  C05\n",
       "6   MMSYN1_0326  F05\n",
       "7   MMSYN1_0077  C06\n",
       "8   MMSYN1_0440  A07\n",
       "9   MMSYN1_0906  C07\n",
       "10  MMSYN1_0220  A08\n",
       "11  MMSYN1_0164  G09\n",
       "12  MMSYN1_0063  G10\n",
       "13  MMSYN1_0314  A11\n",
       "14  MMSYN1_0438  C11"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "good"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%HTML\n",
    "<script src=//cdn.bio.sh/msa/1.0/msa.min.gz.js></script>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_alignments[0].format('clustal')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import HTML\n",
    "    \n",
    "base = HTML('''\n",
    "<script src=//cdn.bio.sh/msa/1.0/msa.min.gz.js></script>\n",
    "\n",
    "<div id=\"alignment\"></div>\n",
    "\n",
    "<script type=\"text/Javascript\">\n",
    "var input = `''' + bad_alignments[0].format('clustal') + '''`;\n",
    "var seqs = msa.io.clustal.parse(input);\n",
    "\n",
    "var opts = {\n",
    "    el: document.getElementById(\"alignment\"),\n",
    "    vis: {\n",
    "      conserv: false,\n",
    "      overviewbox: false\n",
    "    },\n",
    "    // smaller menu for JSBin\n",
    "    menu: \"small\",\n",
    "    bootstrapMenu: false,\n",
    "    seqs: seqs\n",
    "};\n",
    "\n",
    "var m = new msa.msa(opts);\n",
    "m.render();\n",
    "</script>''')we\n",
    "\n",
    "from IPython import display\n",
    "display.display(base)\n",
    "\n",
    "for align in bad_alignments:\n",
    "    code = HTML('''\n",
    "        <script type=\"text/Javascript\">\n",
    "            m.seqs = msa.io.clustal.parse(`{}`);\n",
    "            m.render();\n",
    "        </script>\n",
    "    '''.format(align.format('clustal')))\n",
    "    display.clear_output(wait=True)\n",
    "    display.display((base, code))\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
